unit_tests/log_transfer/test_main_chkpt_info.cpp

```diff
@@ -0,0 +1,800 @@
+{
+  REQUIRE (before.m_start_redo_lsa == after.m_start_redo_lsa);
+  REQUIRE (before.m_snapshot_lsa == after.m_snapshot_lsa);
+
+  REQUIRE (before.m_trans == after.m_trans);
+  REQUIRE (before.m_sysops == after.m_sysops);
+
+  REQUIRE (before.m_has_2pc == after.m_has_2pc);
+}
+
+LOG_2PC_GTRINFO
+test_env_chkpt::generate_2pc_gtrinfo()
+{
+  log_2pc_gtrinfo pc;
+  pc.info_length = rand() % MAX_RAND;
+  pc.info_data = malloc (pc.info_length);
+
+  return pc;
+}
+
+LOG_2PC_COORDINATOR *
+test_env_chkpt::generate_2pc_coordinator()
+{
+  log_2pc_coordinator *pc = new log_2pc_coordinator();
+  pc->num_particps = rand() % MAX_RAND;
+  pc->particp_id_length = rand() % MAX_RAND;
+
+  return pc;
+}
+
+CLIENTIDS
+test_env_chkpt::generate_client (int index)
+{
+  clientids clnt;
+  char number[25];
+
+  clnt.client_type = static_cast<db_client_type> (1);
+  sprintf (number, "client_%d", index);
+  clnt.client_info = (number);
+  sprintf (number, "db_user_%d", index);
+  clnt.db_user = (number);
+  sprintf (number, "database_%d", index);
+  clnt.program_name = (number);
+  sprintf (number, "login_%d", index);
+  clnt.login_name = (number);
+  sprintf (number, "host_%d", index);
+  clnt.host_name = (number);
+  clnt.process_id = rand() % MAX_RAND;
+
+  return clnt;
+}
+
+LOG_TDES *
+test_env_chkpt::generate_tdes (int index)
+{
+  LOG_TDES *tdes = new log_tdes();
+  tdes->trid = index + 1;
+  tdes->tran_index = index + 1;
+
+  tdes->tail_lsa = generate_log_lsa();
+  tdes->isloose_end = std::rand() % 2;
+  tdes->head_lsa = generate_log_lsa();
+  tdes->state    = static_cast<TRAN_STATE> (std::rand() % TRAN_UNACTIVE_UNKNOWN);
+
+  tdes->undo_nxlsa = generate_log_lsa();
+  tdes->posp_nxlsa = generate_log_lsa();
+  tdes->savept_lsa = generate_log_lsa();
+  tdes->tail_topresult_lsa = generate_log_lsa();
+  tdes->rcv.tran_start_postpone_lsa = generate_log_lsa();
+  tdes->wait_msecs = rand() % MAX_RAND;
+  tdes->client_id  = rand() % MAX_RAND;
+  tdes->gtrid      = rand() % MAX_RAND;
+
+  tdes->num_transient_classnames = rand() % MAX_RAND;
+  tdes->num_repl_records         = rand() % MAX_RAND;
+  tdes->cur_repl_record          = rand() % MAX_RAND;
+  tdes->append_repl_recidx       = rand() % MAX_RAND;
+  tdes->fl_mark_repl_recidx      = rand() % MAX_RAND;
+  tdes->repl_insert_lsa          = generate_log_lsa();
+  tdes->repl_update_lsa          = generate_log_lsa();
+
+  tdes->client  = generate_client (index);
+  tdes->gtrinfo = generate_2pc_gtrinfo();
+  tdes->coord   = generate_2pc_coordinator();
+
+  return tdes;
+}
+
+
+void
+test_env_chkpt::generate_tran_table()
+{
+  log_Gl.trantable.num_total_indices = 100;
+
+  int size = log_Gl.trantable.num_total_indices * sizeof (*log_Gl.trantable.all_tdes);
+  log_Gl.trantable.all_tdes = (LOG_TDES **) malloc (size);
+
+  for (int i = 1; i < log_Gl.trantable.num_total_indices; i++)
+    {
+      log_Gl.trantable.all_tdes[i] = generate_tdes (i);
+    }
```

Comment:
There are not top (system) operations generated in generate_tran_table. Generate at least two tdes in transaction table with top operations: one for atomic_sysop_start_lsa not null, one for sysop_start_postpone_lsa not null).

Also generate system worker transaction descriptors with and without top operations.